File: cpp/betterkhmer/test/test_normalize.cpp
#include "betterkhmer.hpp"
#include <cstdio>
#include <cstring>
#include <fstream>
#include <string>

int main(int argc, char** argv) {
    const char* fixtures = argc > 1 ? argv[1] : "../../fixtures";

    // ខ្មែរ
    const std::string khmer = "\xe1\x9e\x81\xe1\x9f\x92\xe1\x9e\x98\xe1\x9f\x82\xe1\x9e\x9a";
    if (betterkhmer::normalize(khmer) != khmer) {
        std::fprintf(stderr, "basic FAILED\n"); return 1;
    }
    std::printf("basic: ok\n");

    char inp_path[4096], exp_path[4096];
    std::snprintf(inp_path, sizeof inp_path, "%s/input.txt",    fixtures);
    std::snprintf(exp_path, sizeof exp_path, "%s/expected.txt", fixtures);

    std::ifstream fi(inp_path), fe(exp_path);
    if (!fi || !fe) { std::fprintf(stderr, "cannot open fixtures\n"); return 1; }

    int total = 0, failures = 0;
    std::string in, ex;
    while (std::getline(fi, in) && std::getline(fe, ex)) {
        std::string result = betterkhmer::normalize(in);
        if (result != ex) {
            failures++;
            if (failures <= 10) std::fprintf(stderr, "[%d] mismatch\n", total);
        }
        total++;
    }

    if (failures > 0) {
        std::fprintf(stderr, "%d/%d fixture failures\n", failures, total);
        return 1;
    }
    std::printf("fixtures: %d ok\n", total);
    return 0;
}

File: zig/betterkhmer/test/test_normalize.zig
const std = @import("std");
const betterkhmer = @import("betterkhmer");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const alloc = gpa.allocator();

    const args = try std.process.argsAlloc(alloc);
    defer std.process.argsFree(alloc, args);
    const fixtures = if (args.len > 1) args[1] else "../../fixtures";

    // basic test — ខ្មែរ
    const khmer = "\xe1\x9e\x81\xe1\x9f\x92\xe1\x9e\x98\xe1\x9f\x82\xe1\x9e\x9a";
    const got = try betterkhmer.normalize(alloc, khmer, "km");
    defer alloc.free(got);
    if (!std.mem.eql(u8, got, khmer)) {
        std.debug.print("basic FAILED\n", .{});
        std.process.exit(1);
    }
    std.debug.print("basic: ok\n", .{});

    // fixture tests
    var inp_buf: [4096]u8 = undefined;
    var exp_buf: [4096]u8 = undefined;
    const inp_path = try std.fmt.bufPrint(&inp_buf, "{s}/input.txt",    .{fixtures});
    const exp_path = try std.fmt.bufPrint(&exp_buf, "{s}/expected.txt", .{fixtures});

    const fi = try std.fs.cwd().openFile(inp_path, .{});
    defer fi.close();
    const fe = try std.fs.cwd().openFile(exp_path, .{});
    defer fe.close();

    var fi_buf = std.io.bufferedReader(fi.reader());
    var fe_buf = std.io.bufferedReader(fe.reader());
    var fi_r = fi_buf.reader();
    var fe_r = fe_buf.reader();

    var total: usize = 0;
    var failures: usize = 0;

    var in_line  = std.ArrayList(u8).init(alloc); defer in_line.deinit();
    var exp_line = std.ArrayList(u8).init(alloc); defer exp_line.deinit();

    while (true) {
        in_line.clearRetainingCapacity();
        exp_line.clearRetainingCapacity();
        fi_r.readUntilDelimiterArrayList(&in_line,  '\n', 65536) catch |e| {
            if (e == error.EndOfStream) break else return e;
        };
        fe_r.readUntilDelimiterArrayList(&exp_line, '\n', 65536) catch |e| {
            if (e == error.EndOfStream) break else return e;
        };

        const result = try betterkhmer.normalize(alloc, in_line.items, "km");
        defer alloc.free(result);

        if (!std.mem.eql(u8, result, exp_line.items)) {
            failures += 1;
            if (failures <= 10)
                std.debug.print("[{}] mismatch\n", .{total});
        }
        total += 1;
    }

    if (failures > 0) {
        std.debug.print("{}/{} fixture failures\n", .{ failures, total });
        std.process.exit(1);
    }
    std.debug.print("fixtures: {} ok\n", .{total});
}
